### PR TITLE
BaSP-TG-6: Get and create projects

### PR DESCRIPTION
**Pull request by LucianaPrat**
Implementation of methods:
-Get project by Id.
-Get list of projects.
-Create a new project.

### CODE REVIEW

File: src/models/Projects.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  endDate: { type: Date, require: true },
+  active: { type: String, require: true },
+  clientName: { type: String, require: true },
+  teamMembers: [{
+    name: { type: String, required: true },
+    rol: { type: String, enum: ['DEV', 'TL', 'QA'] },
+    rate: {
+      type: Number,
+      min: 1,
+      max: 1000,
```
Comment (nicolasblasco) [inline src/models/Projects.js:16]: We shouldn't add validations in the schema

File: src/validations/projects.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        'any.required': 'name is required',
+      }),
+    rol: Joi.string()
+      .required()
+      .messages({
+        'string.empty': 'rol is required',
+        'string.base': 'rol should be DEV, TL or QA',
+        'any.required': 'rol is required',
+      }),
+    rate: Joi.number()
```
Comment (nicolasblasco) [inline src/validations/projects.js:23]: We should add a max and min to the rates number validation. 
Check the documentation here: 

- [validate minimum number](https://joi.dev/api/?v=17.6.1#numberlesslimit)
- [validate maximum number](https://joi.dev/api/?v=17.6.1#numbermaxlimit)

File: src/models/Projects.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+const projectSchema = new mongoose.Schema({
+  name: { type: String, require: true },
+  description: { type: String, require: true },
+  startDate: { type: Date, default: () => Date.now().toString() },
+  endDate: { type: Date, require: true },
+  active: { type: String, require: true },
+  clientName: { type: String, require: true },
+  teamMembers: [{
+    name: { type: String, required: true },
+    rol: { type: String, enum: ['DEV', 'TL', 'QA'] },
```
Comment (nicolasblasco) [inline src/models/Projects.js:12]: We should add the role PM

File: src/validations/projects.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        'string.base': 'clientName sould be only letters',
+        'string.min': 'clientName sould have a minimum length of 2 character',
+        'string.max': 'clientName sould have a maximun length of 150 character',
+        'any.required': 'clientName is required',
+      }),
+    teamMembers: Joi.array()
+      .items(teamMembersValidation)
+      .messages({
+        'string.empty': 'teamMembers is required',
+        'string.base': 'the rol is not valid',
```
Comment (nicolasblasco) [inline src/validations/projects.js:86]: We should specify the roles that can be valid

File: src/validations/projects.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      }),
+  });
+
+  const validation = projectvalidation.validate(req.body, { abortEarly: false });
+  if (validation.error) {
+    return res.status(400).json({
+      message: `There was a validation error: ${validation.error.details[0].message} `,
```
Comment (nicolasblasco) [inline src/validations/projects.js:93]: We should use **validation.errors.details** so we can return more than one error

File: src/controllers/projects.js [hunk 6]
```diff
@@ -1,177 +1,97 @@
...
+      return res.status(400).json({
+        message: `Invalid id: ${req.params.id}`,
+        data: undefined,
+        error: true,
+      });
+    }
+    const project = await Projects.findById(id);
+    if (!project) {
+      return res.status(404).json({
+        msg: `Couldnt find project with id ${id}`,
```
Comment (nicolasblasco) [inline src/controllers/projects.js:48]: Couldn't is mispelled and msg should be message

File: src/controllers/projects.js [hunk 7]
```diff
@@ -1,177 +1,97 @@
...
+        data: undefined,
+        error: true,
       });
+    }
+    return res.status(200).json({
+      msg: 'Project found succesfully',
```
Comment (nicolasblasco) [inline src/controllers/projects.js:54]: msg should be message

File: src/controllers/projects.js [hunk 8]
```diff
@@ -1,177 +1,97 @@
...
+      data: project,
+      error: false,
     });
-  } else {
-    res.status(400).json({
-      success: false,
-      msg: errors,
+  } catch (error) {
+    return res.status(400).json({
+      msg: 'An error ocurred',
```
Comment (nicolasblasco) [inline src/controllers/projects.js:60]: msg should be message and we should return the error in the message. Like this message: error

File: src/controllers/projects.js [hunk 9]
```diff
@@ -1,177 +1,97 @@
...
+      msg: 'Project found succesfully',
+      data: project,
+      error: false,
     });
-  } else {
-    res.status(400).json({
-      success: false,
-      msg: errors,
+  } catch (error) {
+    return res.status(400).json({
```
Comment (nicolasblasco) [inline src/controllers/projects.js:59]: The status code should be 500

File: src/controllers/projects.js [hunk 10]
```diff
@@ -1,177 +1,97 @@
...
+const createProject = async (req, res) => {
+  try {
+    const newProject = new Projects({
+      name: req.body.name,
+      description: req.body.description,
+      startDate: req.body.startDate,
+      endDate: req.body.endDate,
+      active: req.body.active,
+      clientName: req.body.clientName,
+      teamMembers: req.body.teamMembers,
```
Comment (nicolasblasco) [inline src/controllers/projects.js:76]: We could directly do Projects(req.body)

File: src/controllers/projects.js [hunk 11]
```diff
@@ -1,177 +1,97 @@
...
+      data: result,
+      error: false,
     });
-  } else {
-    res.status(404).json({
-      success: false,
-      msg: `There is no project with this id (${projectId})`,
+  } catch (error) {
+    return res.status(500).json({
+      data: undefined,
```
Comment (nicolasblasco) [inline src/controllers/projects.js:87]: We should add message: error

File: src/controllers/projects.js [hunk 12]
```diff
@@ -1,177 +1,97 @@
...
-        res.status(400).json({
-          success: false,
-        });
-        return;
-      }
-      res.status(200).json({
-        succes: true,
-        msg: 'Project deleted successfully',
-      });
+    const result = await newProject.save();
```
Comment (nicolasblasco) [inline src/controllers/projects.js:79]: We could rename this variable to newProjectSaved

**Review by nicolasblasco — CHANGES_REQUESTED** [review]

File: src/controllers/projects.js [hunk 13]
```diff
@@ -1,177 +1,97 @@
...
-    if (!newProject[data]) {
-      errors += err;
+const getProjectById = async (req, res) => {
+  const isValidObjectId = (id) => {
+    if (ObjectId.isValid(id)) {
+      if ((String)(new ObjectId(id)) === id) {
+        return true;
+      }
     }
+    return false;
```
Comment (nicolasblasco) [inline src/controllers/projects.js:34]: This function should go above all the controlllers

**Review by Alfondalix — APPROVED** [review]

**Pull request merged**